contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+        sale_multiplier: Balance,
+        ticket_token: AccountId,
+        reward_token: AccountId,
+    }
+
+    #[derive(Clone, Eq, PartialEq, Debug, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        MissingRole(String),
+        ContractCall(String),
+        PSP22TokenCall(String),
+        MarketplaceEmpty,
+    }
+
+    #[ink(event)]
+    #[derive(Clone, Eq, PartialEq, Debug)]
+    pub struct Buy {
+        #[ink(topic)]
+        pub account_id: AccountId,
+        pub price: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Clone, Eq, PartialEq, Debug)]
+    pub struct Reset;
+
+    impl Error {
+        fn missing_role(role: Role) -> Self {
+            MissingRole(format!("{:?}", role))
+        }
+    }
+
+    impl From<ink_env::Error> for Error {
+        fn from(inner: ink_env::Error) -> Self {
+            Error::ContractCall(format!("{:?}", inner))
+        }
+    }
+
+    impl From<PSP22Error> for Error {
+        fn from(inner: PSP22Error) -> Self {
+            Error::PSP22TokenCall(format!("{:?}", inner))
+        }
+    }
+
+    impl AccessControlled for Marketplace {
+        type ContractError = Error;
+    }
+
+    impl Marketplace {
+        #[ink(constructor)]
+        pub fn new(
+            ticket_token: AccountId,
+            reward_token: AccountId,
+            starting_price: Balance,
+            min_price: Balance,
+            sale_multiplier: Balance,
+            auction_length: BlockNumber,
+        ) -> Self {
+            Self::ensure_role(Self::initializer())
+                .unwrap_or_else(|e| panic!("Failed to initialize the contract {:?}", e));
+
+            Marketplace {
+                ticket_token,
+                reward_token,
+                min_price,
+                sale_multiplier,
+                auction_length,
+                current_start_block: Self::env().block_number(),
+                total_proceeds: starting_price.saturating_div(sale_multiplier),
+                tickets_sold: 1,
+            }
+        }
+
+        #[ink(message)]
+        /// The length of each auction of a single ticket in blocks.
+        ///
+        /// The contract will decrease the price linearly from `average_price() * sale_multiplier()`
+        /// to `min_price()` over this period. The auction doesn't end after the period elapses -
+        /// the ticket remains available for purchase at `min_price()`.
+        pub fn auction_length(&self) -> BlockNumber {
+            self.auction_length
+        }
+
+        #[ink(message)]
+        /// The block at which the auction of the current ticket started.
+        pub fn current_start_block(&self) -> BlockNumber {
+            self.current_start_block
+        }
+
+        #[ink(message)]
+        /// The price the contract would charge when buying at the current block.
+        pub fn price(&self) -> Balance {
+            self.current_price()
```

Comment:
Since `current_price()` can be quite "expensive" and this `price()` method is available in the public API, I would try to cache the `current_price()` computation and reuse it in here.

Reply:
Do you mean make this `pub fn price(&mut self)`? That would make people pay gas for calling this function, wouldn't it?

Reply:
From your other comment, it seems you more so mean pre-computing the parts of the expression that don't depend on `current_block`?

Reply:
I would probably have a field internal to the struct that is set a tuple of block height and the price computed at that height. On every call to `price` I would check if `env::block()` is equal to the cached height and if so return the cached price.

Reply:
Hm, that doesn't answer my question. I'm more worried when that field would be set, not how it would be computed. As far as I understand it's not a good idea to set it in `price` itself?

Reply:
You would set it just where you do it now. If the check returns that `env::block() > cached.block_height` then you would recompute and update the cache.

Reply:
As stated in the https://github.com/Cardinal-Cryptography/aleph-node/pull/613#issuecomment-1243959223, this optimization isn't really optimizing anything (especially that apparently ink! doesn't allow for mutable methonds in the RPC API).